pandas/core/reshape/concat.py

```diff
@@ -437,6 +440,19 @@ def __init__(
             # #1649
             clean_keys = []
             clean_objs = []
+            if is_iterator(keys):
+                keys = list(keys)
+            if is_iterator(objs):
+                objs = list(objs)
+            if len(keys) != len(objs):
+                # GH#43485
+                warnings.warn(
```

Comment:
Is there any code paths below that can be simplified once this deprecation is enforced?

Reply:
No.  ATM we just silently truncate in the zip iteration a few lines down